### PR TITLE
Prison admin api, for get all prison, active deactivate, and create

### PR DESCRIPTION
**Pull request by AledWEvans**
## What does this pull request do?

add's activate, deactivate and create and get all prisons

## What is the intent behind these changes?

to allow the admin api to manage prisons online

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/controller/PrisonConfigController.kt [hunk 1]
```diff
@@ -66,10 +73,40 @@ class PrisonConfigController(
...
+        description = "prison returned",
+      ),
+      ApiResponse(
+        responseCode = "401",
+        description = "Unauthorized to access this endpoint",
+        content = [Content(mediaType = "application/json", schema = Schema(implementation = ErrorResponse::class))],
+      ),
+      ApiResponse(
+        responseCode = "403",
+        description = "Incorrect permissions to view session templates",
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/controller/PrisonConfigController.kt:92]: very minor - description to be updated.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/controller/PrisonConfigController.kt [hunk 2]
```diff
@@ -93,11 +130,114 @@ class PrisonConfigController(
...
+        content = [Content(mediaType = "application/json", schema = Schema(implementation = ErrorResponse::class))],
+      ),
+      ApiResponse(
+        responseCode = "403",
+        description = "Incorrect permissions to activate prison",
+        content = [Content(mediaType = "application/json", schema = Schema(implementation = ErrorResponse::class))],
+      ),
+      ApiResponse(
+        responseCode = "404",
+        description = "Session Template not found",
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/controller/PrisonConfigController.kt:196]: minor point - description to be updated.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/controller/PrisonConfigController.kt [hunk 3]
```diff
@@ -93,11 +130,114 @@ class PrisonConfigController(
...
+        content = [Content(mediaType = "application/json", schema = Schema(implementation = ErrorResponse::class))],
+      ),
+      ApiResponse(
+        responseCode = "403",
+        description = "Incorrect permissions to deactivate prison",
+        content = [Content(mediaType = "application/json", schema = Schema(implementation = ErrorResponse::class))],
+      ),
+      ApiResponse(
+        responseCode = "404",
+        description = "Session Template not found",
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/controller/PrisonConfigController.kt:231]: description

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/PrisonConfigService.kt [hunk 4]
```diff
@@ -39,4 +52,28 @@ class PrisonConfigService(
   fun getSupportedPrisons(): List<String> {
     return prisonRepository.getSupportedPrisons()
   }
+
+  @Transactional
+  fun createPrison(prisonDto: PrisonDto): PrisonDto {
+    val newPrison = Prison(prisonDto.code, prisonDto.active)
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/PrisonConfigService.kt:58]: shall we add a blank validate method that we can populate with rules later to create / activate / deactivate a prison? we could leave it till later as well

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/PrisonConfigService.kt [hunk 5]
```diff
@@ -39,4 +52,28 @@ class PrisonConfigService(
...
+    savedPrison.excludeDates.addAll(excludeDates)
+
+    return mapEntityToDto(savedPrison)
+  }
+
+  @Transactional
+  fun activatePrison(prisonCode: String): PrisonDto {
+    val prisonToUpdate = findPrisonByCode(prisonCode)
+    prisonToUpdate.active = true
+    return mapEntityToDto(prisonToUpdate)
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/PrisonConfigService.kt:70]: do we not need a repository call here.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/PrisonConfigService.kt [hunk 6]
```diff
@@ -39,4 +52,28 @@ class PrisonConfigService(
...
+  }
+
+  @Transactional
+  fun deActivatePrison(prisonCode: String): PrisonDto {
+    val prisonToUpdate = findPrisonByCode(prisonCode)
+    prisonToUpdate.active = false
+    return mapEntityToDto(prisonToUpdate)
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/PrisonConfigService.kt:77]: same as above.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/dto/PrisonDto.kt [hunk 7]
```diff
@@ -10,7 +10,7 @@ data class PrisonDto(
   var code: String,
 
   @Schema(description = "is prison active", example = "true", required = true)
-  var active: Boolean,
+  var active: Boolean = false,
 
   @Schema(description = "exclude dates", required = false)
   var excludeDates: List<LocalDate> = mutableListOf(),
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/dto/PrisonDto.kt:16]: sorry not part of this change but this could be a Set to avoid duplicates.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/controller/PrisonConfigController.kt [hunk 8]
```diff
@@ -93,11 +130,114 @@ class PrisonConfigController(
       ),
     ],
   )
-  fun getPrison(
+  fun getPrisons(): List<PrisonDto> {
+    return prisonConfigService.getPrisons()
+  }
+
+  @PreAuthorize("hasRole('VISIT_SCHEDULER')")
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/controller/PrisonConfigController.kt:137]: do we update the role to an admin role now? Could leave it i till later though.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/PrisonConfigService.kt (hunk 4 above)
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/PrisonConfigService.kt:58]: feature for later on I think! could put a check here to see if it exist mind and if it does it throws an exception

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/PrisonConfigService.kt (hunk 5 above)
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/PrisonConfigService.kt:70]: nope save when exits the transaction scope. (JPA proxy)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/PrisonConfigService.kt (hunk 6 above)
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/PrisonConfigService.kt:77]: comment same as above

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/dto/PrisonDto.kt (hunk 7 above)
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/dto/PrisonDto.kt:16]: good point I will change

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/controller/PrisonConfigController.kt (hunk 8 above)
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/controller/PrisonConfigController.kt:137]: it's another task

**Review by dhirajshettymoj — APPROVED** [review]
LGTM

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/PrisonConfigService.kt (hunk 5 above)
Comment (AledWEvans) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/PrisonConfigService.kt:74]: actually your right

**Review by dhirajshettymoj — APPROVED** [review]
LGTM
